pypeit/spectrographs/keck_kcwi.py

```diff
@@ -1031,46 +1032,164 @@ def get_rawimage(self, raw_file, det):
+            subkrnx = sigmx / ((np.arange(int(10 * sigmx)) - 5 * sigmx) ** 2 + sigmx ** 2)
+            subkrny = sigmy / ((np.arange(int(10 * sigmy)) - 5 * sigmy) ** 2 + sigmy ** 2)
+        else:
+            msgs.error(f"Unknown kernel: {kernel}")
+        kernel = np.outer(subkrnx, subkrny)
+        kernel /= np.sum(kernel)
+        # Make a grid of coordinates
+        specvec, spatvec = np.arange(img.shape[0]), np.arange(img.shape[1])
+        spat, spec = np.meshgrid(spatvec, specvec / (specvec.size - 1))
+        # Convolve the input image (note: most of the time is spent here)
+        # oaconvolve is the fastest option when the kernel is much smaller dimensions than the image
+        # scale_img = (term0 + term1 * spec + term2*spec**2 + term3*spec**3) * signal.fftconvolve(img, kernel, mode='same')
+        scale_img = (term0 + term1 * spec + term2 * spec ** 2 + term3 * spec ** 3) * signal.oaconvolve(img, kernel,
+                                                                                                       mode='same')
+        spl = interpolate.RectBivariateSpline(specvec, spatvec, scale_img, kx=1, ky=1)
+        return spl(zoom * (specvec + shft_spec), zoom * (spatvec + shft_spat))
+
+    def scattered_light(self, frame, offslitmask, tbl, detpad=300):
+        """ Calculate a model of the scattered light of the input frame.
+
+        For KCWI, the main contributor to the scattered light is referred to as the "narcissistic ghost"
+        by Morrissey et al. (2018), ApJ, 864, 93. This scattered light is thought to be a reflection off the
+        detector that travels back through the optical system. Some fraction gets sent back out to space, while
+        the remained comes back through the optical system and a fuzzy version of this is re-imaged onto the
+        detector. The current KCWI scattered light model is designed to account for these effects.
 
         Parameters
         ----------
         frame : `numpy.ndarray`_
             Raw 2D data frame to be used to compute the scattered light.
-        binning : str, `numpy.ndarray`_, tuple
-            Binning of the frame (e.g. '2x1' refers to a binning of 2 in the spectral
-            direction, and a binning of 1 in the spatial direction). For the supported
-            formats, refer to :func:`~pypeit.core.parse.parse_binning`.
+        offslitmask : `numpy.ndarray`_
+            A boolean mask indicating the pixels that are on/off the slit (True = off the slit)
+        tbl : :class:`~pypeit.metadata.PypeItMetaData`_
+            One row of the fitstbl PypeItMetaData that contains metadata about the file being used to
+            optimize the scattered light model parameters.
+        detpad : :obj:`int`_, optional
+            Number of pixels to pad to each of the detector edges to reduce edge effects.
 
         Returns
         -------
         scatt_img : `numpy.ndarray`_
             A 2D image of the scattered light determined from the input frame
```

Comment:
Update to describe the 3 objects returned.

Reply:
Done!